mod-source-record-storage-server/src/main/java/org/folio/services/handlers/match/AbstractMarcMatchEventHandler.java

```diff
@@ -80,22 +89,55 @@ public CompletableFuture<DataImportEventPayload> handle(DataImportEventPayload p
     MatchDetail matchDetail = retrieveMatchDetail(payload);
     if (isValidMatchDetail(matchDetail)) {
       MatchField matchField = prepareMatchField(record, matchDetail);
-      if (matchField.isDefaultField()) {
-        processDefaultMatchField(matchField, payload.getTenant())
-          .onSuccess(recordCollection -> processSucceededResult(recordCollection.getRecords(), payload, future))
-          .onFailure(throwable -> future.completeExceptionally(new MatchingException(throwable)));
-      } else {
-        recordDao.getMatchedRecords(matchField, typeConnection, isNonNullExternalIdRequired(), 0, 2, payload.getTenant())
-          .onSuccess(recordList -> processSucceededResult(recordList, payload, future))
-          .onFailure(throwable -> future.completeExceptionally(new MatchingException(throwable)));
-      }
-    } else {
-      constructError(payload, format(MATCH_DETAIL_IS_NOT_VALID, matchDetail));
-      future.complete(payload);
+      return retrieveMarcRecords(matchField, payload.getTenant())
+        .compose(localMatchedRecords -> {
+          if (isConsortiumAvailable()) {
+            return matchCentralTenantIfNeededAndCombineWithLocalMatchedRecords(payload, matchField, localMatchedRecords);
+          }
+          return Future.succeededFuture(localMatchedRecords);
+        })
+        .compose(recordList -> processSucceededResult(recordList, payload))
+        .recover(throwable -> Future.failedFuture(mapToMatchException(throwable)))
+        .toCompletionStage().toCompletableFuture();
     }
-    return future;
+    constructError(payload, format(MATCH_DETAIL_IS_NOT_VALID, matchDetail));
+    return CompletableFuture.completedFuture(payload);
+  }
+
+  private Future<List<Record>> matchCentralTenantIfNeededAndCombineWithLocalMatchedRecords(DataImportEventPayload payload, MatchField matchField,
+                                                                                           List<Record> recordList) {
+    return consortiumConfigurationCache.get(RestUtil.retrieveOkapiConnectionParams(payload, vertx))
+      .compose(consortiumConfigurationOptional -> {
+        if (consortiumConfigurationOptional.isPresent() && !consortiumConfigurationOptional.get().getCentralTenantId().equals(payload.getTenant())) {
+          LOG.debug("matchCentralTenantIfNeededAndCombineWithLocalMatchedRecords:: Matching on centralTenant with id: {}",
+            consortiumConfigurationOptional.get().getCentralTenantId());
+          return retrieveMarcRecords(matchField, consortiumConfigurationOptional.get().getCentralTenantId())
+            .onSuccess(result -> {
+              if (!result.isEmpty()) {
+                payload.getContext().put(CENTRAL_TENANT_ID, consortiumConfigurationOptional.get().getCentralTenantId());
+              }
+            })
+            .map(centralTenantResult -> Stream.concat(recordList.stream(), centralTenantResult.stream()).toList());
```

Comment:
Is there a tenant identifier on a record? Thinking it may be good to know which records belong to the central tenant or member tenant.

Reply:
We do not have such an identifier